common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala

```diff
@@ -64,100 +64,165 @@ case class ActivationMessage(override val transid: TransactionId,
-      }
-    }
+object CompletionMessage extends DefaultJsonProtocol {
+  def apply(transid: TransactionId, activation: WhiskActivation, invoker: InvokerInstanceId): CompletionMessage = {
+    CompletionMessage(transid, activation.activationId, Some(activation.response.isWhiskError), invoker)
+  }
+  implicit val serdes = jsonFormat4(
+    CompletionMessage.apply(_: TransactionId, _: ActivationId, _: Option[Boolean], _: InvokerInstanceId))
+}
 
-  def parse(msg: String): Try[ResultMessage] = Try(serdes.read(msg.parseJson))
-  implicit val serdes = jsonFormat2(ResultMessage.apply)
+object ResultMessage extends DefaultJsonProtocol {
+  def apply(transid: TransactionId, activation: WhiskActivation): ResultMessage =
+    ResultMessage(transid, Right(activation))
+  implicit private val eitherSerdes = AcknowledegmentMessage.eitherResponse
+  implicit val serdes = jsonFormat2(ResultMessage.apply(_: TransactionId, _: Either[ActivationId, WhiskActivation]))
 }
 
 object AcknowledegmentMessage extends DefaultJsonProtocol {
-  def parse(msg: String): Try[AcknowledegmentMessage] = {
-    Try(serdes.read(msg.parseJson))
+  def parse(msg: String): Try[AcknowledegmentMessage] = Try(serdes.read(msg.parseJson))
+
+  protected[connector] val eitherResponse = new JsonFormat[Either[ActivationId, WhiskActivation]] {
+    def write(either: Either[ActivationId, WhiskActivation]) = either.fold(_.toJson, _.toJson)
+
+    def read(value: JsValue) = value match {
+      case _: JsString =>
+        // per the ActivationId serializer, an activation id is a String even if it only consists of digits
+        Left(value.convertTo[ActivationId])
+
+      case _: JsObject => Right(value.convertTo[WhiskActivation])
+      case _           => deserializationError("could not read ResultMessage")
+    }
   }
 
   implicit val serdes = new RootJsonFormat[AcknowledegmentMessage] {
-    override def write(obj: AcknowledegmentMessage): JsValue = {
-      obj match {
-        case c: CompletionMessage => c.toJson
-        case r: ResultMessage     => r.toJson
-      }
-    }
+    override def write(m: AcknowledegmentMessage): JsValue = m.toJson
 
+    // The field invoker is only part of CombinedCompletionAndResultMessage and CompletionMessage.
+    // If this field is part of the JSON, we try to deserialize into one of these two types,
+    // and otherwise to a ResultMessage. If all conversions fail, an error will be thrown that needs to be handled.
     override def read(json: JsValue): AcknowledegmentMessage = {
-      json.asJsObject
-      // The field invoker is only part of the CompletionMessage. If this field is part of the JSON, we try to convert
-      // it to a CompletionMessage. Otherwise to a ResultMessage.
-      // If both conversions fail, an error will be thrown that needs to be handled.
+      val obj = json.asJsObject
+
+      obj
         .getFields("invoker")
         .headOption
-        .map(_ => json.convertTo[CompletionMessage])
+        .map(_ => {
```

Comment:
From a code maintainability and data modelling perspective, it would be great to also serialise a `type` field that can be used later on to easily de-serialise the JSON. At the same time, this extension increases message size.

I guess the increased size does not really matter whenever a `WhiskActivation` is embedded because a `type` field is pretty small compared to a `WhiskActivation`.

I think the increased size would only matter for the `CompletionMessage`. Today, small JSONs for `CompletionMessage` would look like (146 Byte in compact form):

```
{"transid":"5808a97c269295220a6d8e78508f118b","activationId":"0f3763366aba46a0b763366abae6a0a4","invoker":{"instance":0,"userMemory":"16384 MB"}}
```

When adding a `type` field, we get (166 Byte in compact form):

```
{"type":"completion","transid":"5808a97c269295220a6d8e78508f118b","activationId":"0f3763366aba46a0b763366abae6a0a4","invoker":{"instance":0,"userMemory":"16384 MB"}}
```

This is a size increase of around 14% - which could be compensated by refactoring user memory handling such that it's no more serialised in the `CompletionMessage` - but only in pings.

So I support @chetanmeh's proposal.

Reply:
We can reduce the size bit further by using `{"t":1}` if needed i.e. encode it as enum. However as overhead is not that big we can keep the more readable string form

> refactoring user memory handling such that it's no more serialised in the CompletionMessage - but only in pings. -> `"invoker":{"instance":0,"userMemory":"16384 MB"}`

Would be good to have a issue for this.

Reply:
I considered these - as well making sure the field names were unique so that we can disambiguate with one field lookup. Will think about it again.

Reply:
I'd rather not change the JSON represntation to include the type. I think it's not necessary at this point but I did simplify the implementation to something more compact.